File: python/sort/merge_sort.py
def merge_sort(array: list[int]) -> list[int]:
    if len(array) == 1:
        return array

    left_half = array[:len(array) // 2]
    right_half = array[len(array) // 2:]
    sorted_left_half = merge_sort(left_half)
    sorted_right_half = merge_sort(right_half)
    return merge(sorted_left_half, sorted_right_half)


def merge(left_array: list[int], right_array: list[int]) -> list[int]:
    left_idx = 0
    right_idx = 0
    result = []

    while True:
        if left_array[left_idx] <= right_array[right_idx]:
            result.append(left_array[left_idx])
            left_idx += 1
            if left_idx == len(left_array):
                result.extend(right_array[right_idx:])
                break
        else:
            result.append(right_array[right_idx])
            right_idx += 1
            if right_idx == len(right_array):
                result.extend(left_array[left_idx:])
                break

    return result

File: go/sort/mergeSort.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	values := readInput()
	result := mergeSort(values)
	fmt.Println(result)
}


func mergeSort(values []int) []int {
	if len(values) <= 1 {
		return values
	}
	left, right := values[:len(values)/2], values[len(values)/2:]
	sortedLeft := mergeSort(left)
	sortedRight := mergeSort(right)
	result := merge(sortedLeft, sortedRight)
	return result
}

func merge(left, right []int) []int {
	var result []int
	var leftIdx, rightIdx int
	for {
		if left[leftIdx] < right[rightIdx] {
			result = append(result, left[leftIdx])
			leftIdx++
			if leftIdx == len(left) {
				result = append(result, right[rightIdx:]...)
				break
			}
		} else {
			result = append(result, right[rightIdx])
			rightIdx++
			if rightIdx == len(right) {
				result = append(result, left[leftIdx:]...)
				break
			}
		}
	}
	return result
}

func readInput() []int {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	scanner.Scan()
	size, _ := strconv.Atoi(scanner.Text())
	values := make([]int, size)
	for idx := range size {
		scanner.Scan()
		values[idx], _ = strconv.Atoi(scanner.Text())
	}
	return values
}
